Mathlib/Data/Fin/Basic.lean

```diff
@@ -1089,8 +1162,125 @@ section Pred
 
 #align fin.nat_add_sub_nat_cast Fin.natAdd_subNat_castₓ
 
+theorem pred_one' [NeZero n] (h := (zero_ne_one' (n := n)).symm):
+    Fin.pred (1 : Fin (n + 1)) h = 0 :=
+    by simp_rw [Fin.ext_iff, coe_pred, val_one', val_zero',
+      tsub_eq_zero_iff_le, Nat.mod_le]
+
+theorem pred_last (h := last_pos'.ne') :
+    pred (last (n + 1)) h = last n := by simp_rw [← succ_last, pred_succ]
+
+theorem pred_lt_iff {i : Fin (n + 1)} (hi : i ≠ 0) : pred i hi < j ↔ i < succ j := by
+  rw [← succ_lt_succ_iff, succ_pred]
+theorem lt_pred_iff {i : Fin (n + 1)} (hi : i ≠ 0) : j < pred i hi ↔ succ j < i := by
+  rw [← succ_lt_succ_iff, succ_pred]
+theorem pred_le_iff {i : Fin (n + 1)} (hi : i ≠ 0) : pred i hi ≤ j ↔ i ≤ succ j := by
+  rw [← succ_le_succ_iff, succ_pred]
+theorem le_pred_iff {i : Fin (n + 1)} (hi : i ≠ 0) : j ≤ pred i hi ↔ succ j ≤ i := by
+  rw [← succ_le_succ_iff, succ_pred]
+
+/- Move this somewhere. Not original. -/
+theorem castSucc_pred_eq_pred_castSucc {a : Fin (n + 1)} (ha : a ≠ 0)
+    (ha' := a.castSucc_ne_zero_iff.mpr ha) :
+    castSucc (a.pred ha) = (castSucc a).pred ha' := rfl
+#align fin.cast_succ_pred_eq_pred_cast_succ Fin.castSucc_pred_eq_pred_castSucc
+
 end Pred
 
+section CastPred
+
+/-- TODO -/
+
+@[inline] def castPred (i : Fin (n + 1)) (h : i ≠ last _) : Fin n := ⟨i, val_lt_last h⟩
+#align fin.cast_pred Fin.castPred
```

Comment:
From what I can tell, this is not just refactoring `castPred`, but changing it's meaning from mapping `last (n+1) => last n` to restricting the domain with the `i /= last _` assumption.
I don't think this is desirable from a backwards compatibility POV. Maybe we rename this definition to `castPredOfNe`, and define `castPred` in terms of it and `if-then-else`?

Reply:
No - I'm making the change to make it consistent with pred. If pred was instead predOfNe I would agree, but the current state of affairs where they are inconsistent is not desirable, and I think this is better.

Note that the current castPred is exactly predAbove 0.

Reply:
It's confusing that pred and castPred currently work differently, and the only reason for this that I can see is that someone did a so-so job in it in the first place.